### ISSUE TITLE
Perf: Reduce per-iteration allocations in training loop

### ISSUE THREAD

**Issue by stservice** [stSoftwareAU/NEAT-TS#1540]
## Summary

Several allocations in `src/architecture/Training.ts` occur per training iteration that could be avoided:

### 1. BackPropagation config spread+freeze per iteration
```ts
const iterationConfig = createBackPropagationConfig({
  ...backPropConfig,
  generations: backPropConfig.generations + iteration,
  learningRate: currentLearningRate,
});
```
Creates a new frozen object every iteration. Only `generations` and `learningRate` change.

### 2. Int32Array allocations in trainDirBinary
```ts
const tmpIndexes = new Int32Array(fileRecords);
// ...
const selectedIndexes = new Int32Array(actualLen);
```
Two typed arrays allocated per file per iteration for sample index shuffling.

### 3. SparseConfig rebuild
`new SparseConfig(...)` calls `chooseNeurons()` (random selection) and `calculatePathsToOutput()` (graph traversal) per iteration, even when neuron errors are empty or unchanged.

## What Needs to Be Done

1. **BackPropagation config**: Accept a mutable iteration config struct with updatable `generations` and `learningRate` fields, rather than creating a frozen copy each iteration.
2. **Index arrays**: Pre-allocate a scratch `Int32Array` at the maximum expected size and reuse it across iterations. Use the `BufferPool` pattern if applicable.
3. **SparseConfig**: Skip rebuild when `neuronErrors` is `undefined` or empty and `sparseRatio === 1`. Add an early return guard before the `SparseConfig` constructor call.

## Acceptance Criteria

- [ ] No new object allocation per iteration for backprop config (mutate in place or reuse)
- [ ] Index arrays are reused across iterations where possible
- [ ] `SparseConfig` is not rebuilt when neuron errors are empty
- [ ] All training tests pass
- [ ] Benchmark training throughput before and after

Part of #1533
